### PR TITLE
Feature/query command

### PR DESCRIPTION
**Pull request by Femi3211**

### CODE REVIEW

File: cli/src/main/java/com/adaptivescale/rosetta/cli/model/Config.java [hunk 1]
```diff
@@ -1,13 +1,20 @@
...
 import com.adaptivescale.rosetta.common.models.input.Connection;
+import com.fasterxml.jackson.annotation.JsonProperty;
 
 import java.util.List;
 import java.util.Optional;
 
 public class Config {
     private List<Connection> connections;
 
+    @JsonProperty("OpenAI_API_KEY")
```
Comment (nbesimi) [inline cli/src/main/java/com/adaptivescale/rosetta/cli/model/Config.java:12]: Maybe it is better if we name this: `openai_api_key`
and for the Java part use `openAIApiKey`

File: cli/src/main/java/com/adaptivescale/rosetta/cli/model/Config.java [hunk 2]
```diff
@@ -1,13 +1,20 @@
...
+    private String OpenAI_API_KEY;
+
+    @JsonProperty("OpenAI_Model")
```
Comment (nbesimi) [inline cli/src/main/java/com/adaptivescale/rosetta/cli/model/Config.java:15]: Maybe change this to `openai_model` and the Java part to `openAIModel

File: queryhelper/src/main/java/queryhelper/service/AIService.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+import java.sql.Driver;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+
+public class AIService {
+    private final static String AI_MODEL = "gpt-3.5-turbo";
```
Comment (nbesimi) [inline queryhelper/src/main/java/queryhelper/service/AIService.java:27]: We might need to write in README.md that the default model is `gpt-3.5-turbo`

File: queryhelper/src/main/java/queryhelper/service/AIService.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        long startTime;
+        long endTime;
+        double elapsedTime;
+        String query;
+        String aiOutputStr;
+        String prompt;
+
+        OpenAiChatModel.OpenAiChatModelBuilder model = OpenAiChatModel
+                .builder()
+                .temperature(0.1)
```
Comment (nbesimi) [inline queryhelper/src/main/java/queryhelper/service/AIService.java:43]: Any reason why we are not using temperature value **0.0**? I have seen that **0.1** works great so far.

File: queryhelper/src/main/java/queryhelper/service/AIService.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+        try {
+            QueryRequest aiOutputObj = gson.fromJson(aiOutputStr, QueryRequest.class);
+            query = aiOutputObj.getQuery();
+        } catch (Exception e) {
+            return ErrorUtils.invalidResponseError(e);
+        }
+
+        try {
+            net.sf.jsqlparser.statement.Statement statement = CCJSqlParserUtil.parse(query);
```
Comment (nbesimi) [inline queryhelper/src/main/java/queryhelper/service/AIService.java:74]: Can we add some unit tests around the `CCJSqlParserUtil.parse(query)`

File: README.md [hunk 6]
```diff
@@ -653,6 +653,51 @@ Parameter | Description
 --pyspark | Generates the Spark SQL file.
 --scala | Generates the Scala SQL file.
 
+#### query
```
Comment (nbesimi) [inline README.md:656]: The query command allows you to use natural language commands to query your databases, transforming these commands into SQL SELECT statements. By leveraging the capabilities of AI and LLMs, specifically OpenAI models, it interprets user queries and generates the corresponding SQL queries. For effective use of this command, users need to provide their OpenAI API Key and specify the OpenAI model to be utilized. The output will be written to a CSV file. The max number of rows that will be returned is 200. You can overwrite this value, or remove completely the limit.

File: README.md [hunk 7]
```diff
@@ -653,6 +653,51 @@ Parameter | Description
...
+The query command helps you query your databases using natural language commands, converting them into SQL SELECT statements. This command leverages the capabilities of OpenAI models to interpret user queries and generate corresponding SQL queries. To utilize this command effectively, users must provide their OpenAI API Key and specify the OpenAI model to be used, output will be written in a CSV file.   
+
+    rosetta [-c, --config CONFIG_FILE] query [-h, --help] [-s, --source CONNECTION_NAME] [-q, --query Natural language QUERY]
```
Comment (nbesimi) [inline README.md:659]:
```
rosetta [-c, --config CONFIG_FILE] query [-h, --help] [-s, --source CONNECTION_NAME] [-q, --query "Natural language QUERY"]
```

File: README.md [hunk 8]
```diff
@@ -653,6 +653,51 @@ Parameter | Description
...
+-q --query Natural language QUERY  | pecifies the natural language query to be transformed into an SQL SELECT statement.
+-l --limit Response Row limit (Optional) | Limits the number of rows in the generated CSV file. If not specified, the default limit is set to 200 rows.
+--no-limit Response Row limit (Optional) | Specifies that there should be no limit on the number of rows in the generated CSV file.
+
+
+**Example** (Setting the key and model) :
+
+(Config file)
+```
+OpenAI_API_KEY: "sk-abcdefghijklmno1234567890"
```
Comment (nbesimi) [inline README.md:675]: This will get updated as result of the comments bellow.

File: README.md [hunk 9]
```diff
@@ -653,6 +653,51 @@ Parameter | Description
...
+OpenAI_Model: "gpt-4"
```
Comment (nbesimi) [inline README.md:676]: Same for this

File: README.md (hunk 6 above)
Comment (nbesimi) [reply README.md:656]: Here as a last sentence we can add that:

**The default openai model that is used is gpt-3.5-turbo**

File: cli/src/main/java/com/adaptivescale/rosetta/cli/Cli.java [hunk 10]
```diff
@@ -517,4 +520,50 @@ public FileNameAndDatabasePair(String key, Database value) {
...
+
+        DDL modelDDL = DDLFactory.ddlForDatabaseType(source.getDbType());
+        String DDL = modelDDL.createDatabase(db, false);
+
+        QueryRequest queryRequest = new QueryRequest();
+        queryRequest.setQuery(userQueryRequest);
+        String timestamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
+        String fileName = userQueryRequest.replaceAll("\\s+", "_") + "_" + timestamp + ".csv";
+
+        Path csvFilePath = dataDirectory.resolve(fileName);
```
Comment (jetroni) [inline cli/src/main/java/com/adaptivescale/rosetta/cli/Cli.java:562]: Maybe making a prep function in AIService?

File: cli/src/main/java/com/adaptivescale/rosetta/cli/model/Config.java (hunk 1 above)
Comment (Femi3211) [reply cli/src/main/java/com/adaptivescale/rosetta/cli/model/Config.java:12]: Yes i agree, this alongside with the model will be renamed

File: queryhelper/src/main/java/queryhelper/service/AIService.java (hunk 4 above)
Comment (Femi3211) [reply queryhelper/src/main/java/queryhelper/service/AIService.java:43]: As far as making the model more deterministic, this needs to be tested, with the temperature 0.1 the prompt has been providing very solid responses, tweaking the prompt may be something that requires more testing, but as it is right now we are not facing any problems

**Pull request merged**